atom/browser/api/save_page_handler.cc

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2015 GitHub, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "atom/browser/api/save_page_handler.h"
+
+#include <string>
+
+#include "atom/browser/atom_browser_context.h"
+#include "atom/browser/atom_download_manager_delegate.h"
+#include "atom/common/native_mate_converters/download_item_converter.h"
+#include "base/callback.h"
+#include "base/files/file_path.h"
+#include "content/public/browser/web_contents.h"
+
+namespace atom {
+
+namespace api {
+
+SavePageHandler::SavePageHandler(content::WebContents* web_contents,
+                                 const SavePageCallback& callback)
+    : id_(0),
+      web_contents_(web_contents),
+      callback_(callback) {
+}
+
+SavePageHandler::~SavePageHandler() {
+  LOG(ERROR) << "~SavePageHandler";
+}
+
+void SavePageHandler::OnDownloadCreated(content::DownloadManager* manager,
+                                        content::DownloadItem* item) {
+  if (id_ == item->GetId()) {
+    GetDownloadManager()->RemoveObserver(this);
+    item->AddObserver(this);
+  }
+}
+
+bool SavePageHandler::Handle(const base::FilePath& full_path,
+                             const content::SavePageType& save_type) {
+  auto delegate = static_cast<AtomDownloadManagerDelegate*>(
+      GetDownloadManager()->GetDelegate());
+  GetDownloadManager()->AddObserver(this);
+
+  id_ = delegate->NextId();
+  bool result = web_contents_->SavePage(full_path,
```

Comment:
Hmm the SavePackage's API actually assumes the request is started by WebContents, otherwise the request would be cancelled:
https://code.google.com/p/chromium/codesearch#chromium/src/content/browser/download/save_file_manager.cc&sq=package:chromium&type=cs&l=180

Sorry for the wrong direction.